### PR TITLE
Support OpenTelemetry Metrics

### PR DESCRIPTION
**Pull request by kmssap**
Signed-off-by: Kai M. Sternad <kai.sternad@sap.com>
Co-authored-by: Tomas Longo <tomas.longo@sap.com>
Co-authored-by: Karsten Schnitter <k.schnitter@sap.com>

### Description
This change introduces support for OpenTelemetry metrics. It adds two plugins, one for receiving Metrics via gRPC and another one for processing and mapping the raw metrics records to JSON records that can be ingested into OpenSearch.

#### Caveats: 
The metrics plugin is structured similarly to the already existing OpenTelemetry tracing plugin and therefore has a number of similarities. Specifically otel-metrics-source basically copies most of otel-trace-source and should ideally be de-duplicated.

Also, in order to be able to use the latest OpenTelemetry protocol changes we upgraded the protocol version from 1.0.1-alpha to 1.7.1-alpha. This may have ramifications for the other otel-related plugins.

### CODE REVIEW

**Comment by KarstenSchnitter** [comment]
Addresses #242

**Comment by dlvenable** [comment]
Thank you for your contribution! This is a great feature for Data Prepper and we are excited to merge in your work. I will prioritize reviewing this PR.

We plan to merge this into `main` in a timely manner. We have been wrapping up the 1.3 release which we have scheduled for Mar 21. I think it would be great to have it in that release, but I'm unsure at the moment. Regardless, I expect to merge this into `main` before that date.

**Comment by KarstenSchnitter** [comment]
This PR currently supports four of the five metric types:

- Sums
- Gauges
- Histograms
- Summaries

The missing ExponentialHistograms can be added, when agreement on the data structure of histograms is reached. We decided for the provided data model to enable easy aggregations in the UI. 

Another missing piece is Exemplars. We envision them as arrays to a metrics document. This makes queries on the exemplars harder, but we see them as auxiliary information not as a main value. 

We are very open to feedback about the missing pieces as well as on the provided PR.

**Comment by kmssap** [comment]
There is a number of issues with outdated, vulnerable armeria (transitive dependency) and protbuf libraries in other projects. We bumped the dependencies in the metrics plugins, but there ist at least one other that is currently blocking the build. We noticed there are other PRs that address this. What would be the best way to proceed ?

**Comment by dlvenable** [comment]
> There is a number of issues with outdated, vulnerable armeria (transitive dependency) and protbuf libraries in other projects. We bumped the dependencies in the metrics plugins, but there ist at least one other that is currently blocking the build. We noticed there are other PRs that address this. What would be the best way to proceed ?

@kmssap, We have hit a similar issue with the WhiteSource Security Check in some of our PRs. If you branched before a library was fixed, then your branch has an old dependency and WhiteSource detects the regression. Normally, we fix this by rebasing onto `main` and then force pushing. It appears you rebased. Correct?

We have been having trouble updating Armeria. I think the best approach for you to proceed regarding dependencies is to use the exact same versions as are used in `otel-trace-source` in main.

https://github.com/opensearch-project/data-prepper/blob/94ca5777bc74f1e78c5ec279ce40fd1c20f8e111/data-prepper-plugins/otel-trace-source/build.gradle#L21-L23

If you continue to get WhiteSource errors, we can manually inspect the build.gradle and I believe we can bypass the failing check if needed.

File: data-prepper-plugins/otel-metrics-raw-processor/build.gradle [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/*
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-raw-processor/build.gradle:1]: You should probably be using the following license header here and in all the source files.

```
/*
  * Copyright OpenSearch Contributors
  * SPDX-License-Identifier: Apache-2.0
  */
```

The longer form is only for files modified from other projects. Data Prepper source files mistakenly used these and now we've updated them. If you have any source files modified from projects outside of Data Prepper you can use the longer header. Otherwise, please update to the short header.

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricSourceConfig.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    static final String SSL_KEY_CERT_FILE = "sslKeyCertChainFile";
+    static final String SSL_KEY_FILE = "sslKeyFile";
+    static final String ACM_CERT_ARN = "acmCertificateArn";
+    static final String ACM_PRIVATE_KEY_PASSWORD = "acmPrivateKeyPassword";
+    static final String AWS_REGION = "awsRegion";
+    static final String THREAD_COUNT = "thread_count";
+    static final String MAX_CONNECTION_COUNT = "max_connection_count";
+    static final String ENABLE_UNFRAMED_REQUESTS = "unframed_requests";
+    static final int DEFAULT_REQUEST_TIMEOUT_MS = 10000;
+    static final int DEFAULT_PORT = 21890;
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricSourceConfig.java:35]: This is the default port that the Trace source uses. I think these should be default values. Otherwise, if you have two pipelines - one for trace and metrics - Data Prepper will fail to start with defaults.

Perhaps `21891` would be a good port number.

File: data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawSummary.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    private final Map<String, Object> attributes;
+    private final String kind;
+    private final String unit;
+    private final String name;
+    private final String description;
+    private final String startTime;
+    private final String time;
+    private final String serviceName;
+    private final int quantileValuesCount;
+    private final List<SummaryDataPoint.ValueAtQuantile> quantileValuesList;
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawSummary.java:29]: Perhaps we should rename this to `quantileValues`. The Trace models don't have a `List` suffix on property names.

File: data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        private final double upperBound;
+        private final long numberOfObservations;
+
+        public Bucket(double lowerBound, double upperBound, long numberOfObservations) {
+            this.lowerBound = lowerBound;
+            this.upperBound = upperBound;
+            this.numberOfObservations = numberOfObservations;
+        }
+
+        @JsonProperty("lo")
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java:118]: Is there a particular reason you used `lo`, `hi`, and `cnt`? Our other models all use whole words and I think it would be good to keep that convention. Also, I believe the OTel documentation seems to refer to these as `min`, `max`, and `count`.

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricSourceConfig.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+ *  GitHub history for details.
+ */
+
+package com.amazon.dataprepper.plugins.source.otelmetrics;
+
+import com.amazon.dataprepper.model.configuration.PluginSetting;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class OTelMetricSourceConfig {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricSourceConfig.java:18]: Data Prepper now supports POJO models for plugin configurations using Jackson serialization and it supports Java bean validation. This isn't a mandatory change since it can be made without a major impact later on. But, I did want to mention it.

You can see how we have it working for Traces now:

https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/otel-trace-source/src/main/java/com/amazon/dataprepper/plugins/source/oteltrace/OTelTraceSourceConfig.java

File: data-prepper-plugins/otel-metrics-source/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+## Usages
+Example `.yaml` configuration:
+```
+source:
+    - otel_metrics_source:
+```
+
+## Configurations
+
+* port(Optional) => An `int` represents the port Otel metrics source is running on. Default is ```21890```.
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-source/README.md:17]: Please update the default port value based on my other comment.

File: data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/OTelMetricsStringProcessor.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@DataPrepperPlugin(name = "otel_metrics_raw_processor", pluginType = Processor.class)
+public class OTelMetricsStringProcessor extends AbstractProcessor<Record<ExportMetricsServiceRequest>, Record<String>> {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/OTelMetricsStringProcessor.java:44]: This class is very significant and should have a unit test.

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+
+@DataPrepperPlugin(name = "otel_metrics_source", pluginType = Source.class)
+public class OTelMetricsSource implements Source<Record<ExportMetricsServiceRequest>> {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java:40]: This class is very significant and should have a unit test. I think you could base it off of the `OTelTraceSourceTest` class.

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsGrpcService.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceRequest;
+import io.opentelemetry.proto.collector.metrics.v1.ExportMetricsServiceResponse;
+import io.opentelemetry.proto.collector.metrics.v1.MetricsServiceGrpc;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeoutException;
+
+public class OTelMetricsGrpcService extends MetricsServiceGrpc.MetricsServiceImplBase {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsGrpcService.java:30]: This class is very significant and should have a unit test. I think you could base it off of the `OTelTraceGrpcServiceTest` class.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks for this contribution! I did make some inline comments. I also have some higher level thoughts and question here.

1. What is your use-case for metric collection? It appears that the current approach is not performing any aggregation and thus will produce lots of documents in OpenSearch. Can you share some about how this is working for your team and if there are any issues with the sheer volume of data going into OpenSearch?
2. I made a few comments about the new `Event` model. We are currently having to support a migration path to the new approach. Previously, Data Prepper placed the `ExportTraceServiceRequest` in the Buffer. Then we split it into `String` objects for each Span. In the upcoming change the OTel Trace Source will write `Span` objects directly into the Buffer. I think this would be a good thing to consider with the addition of metric ingestion. Do you have any thoughts on this approach?
3. You mentioned that you duplicated the `otel_trace_source`. I think this is fine for now. I do think we could later find a way to combine these sources into one.
4. I believe that supporting some metric types is better than supporting none at all. So I'm fine with accepting this work without support for Exemplars.

**Comment by KarstenSchnitter** [comment]
@dlvenable we do not envision OpenSearch as a primary metrics store. But we want to provide a bridge between metrics and logs for more elaborate observability. We did some testing with our approach. Our main focus was to enable proper visualisation and alerting capabilities in the UI. We created some Vega visualisations to verify our approach. Since OpenSearch has very limited support for aggregated data such as histograms, we opted for our solution with a higher document number. OpenSearch proved to be quite capable in handling the metrics volume we tested. In particular it shines when analysing high cardinality field values, which are problematic for most specialised Time-Series Databases. The much higher storage volume is for sure the downside. In the end we want to provide support to enrich our Logging service with additional metrics, that can easily be correlated with each other.

**Comment by kmssap** [comment]
@dlvenable Thanks for your detailed review and the hints regarding the Whitesource Security Check. We'll address the review comments in the following commits.

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricSourceConfig.java (hunk 2 above)
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricSourceConfig.java]: Done

File: data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java (hunk 4 above)
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java:118]: The reasoning behind this decision was to save some space by using short names. We will follow your suggestion and the OTel documentation

**Comment by dlvenable** [comment]
@KarstenSchnitter , Thank you for elaborating on your use-case and what you've experienced. Are you already using this in a production environment? Also, do you configure your metrics providers (upstream from Data Prepper) to supply aggregate metrics at a certain interval? If so, what have you been using?

@kmssap , Thank you for addressing the PR comments. Regarding the `Event` comments, I'd be happy to have further discussion if you have any concerns about these changes.

File: data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java (hunk 4 above)
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java]: Done, it is now "upperBound", "lowerBound" and "count". I couldn't find "min" and "max" in the spec, the spec appears to have no explicit terminology for bucket bounds. Or did I miss something ?

File: data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawSummary.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawSummary.java]: Done

Comment (kmssap) [reply data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawSummary.java]: Done

**Comment by KarstenSchnitter** [comment]
@dlvenable The exact approach provided by this PR is not yet used in production. We use a different approach as a bridging technology until this change becomes available. 

For this we use the telegraf metrics agent and its http json output to ingest metrics into OpenSearch. We apply some processing on the json with Fluentd. Metrics are indexed in their own index. We use a strict retention scheme via ILM to restrict retention to 7-30 days, depending on data volume and requirements. 

Telegraf itself can be used to scrape Prometheus endpoints, either from an existing Prometheus instance. In that case, a downsampling and selection of the overall data contained in that store is highly recommended. For smaller use-cases Telegraf can also scrape applications directly and also collect "environmental" data, such as cpu or memory consumption. 

This approach has shown good performance and stability during testing with approx 30 million metric events per day. Different aggregation intervals 10-120s were used in this approach depending on the data source. This allowed us to validate  our approach to the data modelling we propose in this PR. We still need to repeat this testing using the DataPrepper plugin of this PR though.

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsGrpcService.java (hunk 9 above)
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsGrpcService.java:24]: Done

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java (hunk 8 above)
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java:42]: Done

[the remark this answers is not in the document]
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java:42]: @dlvenable Thanks for the suggestion. We have addressed most of the other review comments and are now looking into migrating this PR towards the new Event model.

[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java:42]: Thank you @kmssap ! One idea I have on this is that we can keep the Source the way it is. And just update the `otel_metrics_raw_processor` to output `Record<Event>`. This would probably be the most straightforward approach to use Events.

Thus we'd have:

```
@DataPrepperPlugin(name = "otel_metrics_raw_processor", pluginType = Processor.class)
public class OTelMetricsStringProcessor extends AbstractProcessor<Record<ExportMetricsServiceRequest>, Record<Event>> {
```

It also would make sense to drop the "String" from the name and just name the processor: `OTelMetricsRawProcessor`

File: data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java [hunk 10]
```diff
@@ -107,27 +107,24 @@ public String toJson() throws JsonProcessingException {
...
-        public Bucket(double lowerBound, double upperBound, long numberOfObservations) {
+        public Bucket(double lowerBound, double upperBound, long count) {
             this.lowerBound = lowerBound;
             this.upperBound = upperBound;
-            this.numberOfObservations = numberOfObservations;
+            this.count = count;
         }
 
-        @JsonProperty("lo")
         public double getLowerBound() {
```
Comment (dlvenable) [inline data-prepper-plugins/otel-metrics-raw-processor/src/main/java/com/amazon/dataprepper/plugins/processor/otelmetrics/model/RawHistogram.java:118]: Thanks for renaming `count`. Why did you rename `lo` and `hi` to `lowerBound` and `upperBound`? I believe OTel uses `min` and `max` and these are both smaller for the OpenSearch documents.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks! I made a couple new inline comments. But, I think this is very close to completion.

File: data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (kmssap) [reply data-prepper-plugins/otel-metrics-source/src/main/java/com/amazon/dataprepper/plugins/source/otelmetrics/OTelMetricsSource.java:42]: @dlvenable Yes this sounds good. What do you think about subclassing JacksonEvent with a new class JacksonMetric and changing the signature of the Processor to output Record<Event> (or Record<Metric>, using a new Interface that extends Event). The input would remain Record<ExportMetricsServiceRequest> for the time being until #546 is merged to main. After this the mapping logic could be migrated to OtelProtoCodec so that the source would also emit Events exclusively.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Histogram.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+     * @since 1.4
+     */
+    String getAggregationTemporality();
+
+    /**
+     * Gets the actual buckets for a histogram
+     * @return the buckets
+     * @since 1.4
+     */
+    List<JacksonHistogram.Bucket> getBuckets();
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Histogram.java:50]: This interface shouldn't have any reference to the Jackson implementations. We can follow the pattern that the trace model uses for `TraceGroupFields`. It has an interface for `TraceGroupFields` and then an implementation `DefaultTraceGroupFields`.

Also, Java supports static inner classes within interfaces. If you do not believe there is much value in using an interface here, we could move the `Bucket` class into `Histogram` from `JacksonHistogram`.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Summary.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+ * @since 1.4
+ */
+public interface Summary extends Metric {
+
+    /**
+     * Gets the quantiles for a summary
+     * @return the quantiles
+     * @since 1.4
+     */
+    List<JacksonSummary.SummaryQuantile> getQuantiles();
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Summary.java:21]: Similar to my other comment for `Histogram.Bucket`, this shouldn't reference the Jackson class. Please take a similar approach for `SummaryQuantile` as you do for `Bucket`.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/ParameterValidator.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+package com.amazon.dataprepper.model.metric;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+public class ParameterValidator {
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/ParameterValidator.java:15]: It seems this class is only required for the `Jackson` classes in this package. Please make it package protected so that it does not get used beyond there. In Java, you just remove the `public` modifier before class.

```
class ParameterValidator
```

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/JacksonHistogram.java [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+
+        private void checkAndSetDefaultValues() {
+            data.computeIfAbsent(ATTRIBUTES_KEY, k -> new HashMap<>());
+        }
+
+    }
+
+    public static class Bucket {
+
+        private final double lowerBound;
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/JacksonHistogram.java:182]: Please rename to `min` to keep the OTel terminology and reduce size.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/JacksonHistogram.java [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+        private final double upperBound;
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/JacksonHistogram.java:183]: Please rename to `max`.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks for making this change! It looks very good. I have a few relatively small comments. After addressing those, we should be good to go.

**Comment by dlvenable** [comment]
One other comment: The DCO check is failing. Please see the bottom of this page: https://github.com/opensearch-project/data-prepper/pull/1154/checks?check_run_id=5650732139

It appears that it is expecting no middle initial, but got one.

I believe the DCO check will check any `Signed-off-by`. So you could add a new one to those commits and it should work.

**Comment by dlvenable** [comment]
@kaisternad , You can ignore my previous comment about the DCO. From what I have found, it is still acceptable. It is just our DCO check automation is strict.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Histogram.java (hunk 11 above)
Comment (kmssap) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Histogram.java:50]: Thanks @dlvenable ! We missed this one, also the Quantile. We'll address this and the other review comments in the next commits

**Comment by dlvenable** [comment]
It looks like some of the tests are failing as well.

One I noticed is that the code coverage for the `data-prepper-api` project has gone down:

> [ant:jacocoReport] Rule violated for bundle data-prepper-api: instructions covered ratio is 0.7, but expected minimum is 1.0

This project has a 100% coverage requirement. A few of the projects have this requirement, and this is one of them.

This build is generally consistent: `Data Prepper Java CI with Gradle`. You should be able to get this one passing.

I'm not sure about the other errors. We have some flaky tests and that might be the cause. I can re-run these if they continue to fail after the main Gradle build is running.

**Comment by kmssap** [comment]
@dlvenable Thanks, yes we are aware of the decreased api coverage. This is due to the newly introduced builders. We are about to address this as well.

File: data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Histogram.java (hunk 11 above)
Comment (kmssap) [reply data-prepper-api/src/main/java/com/amazon/dataprepper/model/metric/Histogram.java]: We added a Bucket interface and a default implementation, as suggested

**Comment by dlvenable** [comment]
@kaisternad , Generally the code looks good.

The Gradle build is currently failing due to checkstyles errors. The Data Prepper project doesn't use wildcard imports, so checkstyle is failing.

You should be able to see the output here:
https://github.com/opensearch-project/data-prepper/runs/5697654738?check_suite_focus=true

You can run the full Gradle build locally by running:

```
./gradlew build
```

It will run the unit tests, integration tests, and checkstyle.

I think we just need to get the build running and then we should be good to merge. Is there anything else you need to add?

**Review by dlvenable — CHANGES_REQUESTED** [review]
Address failing builds

**Review by dlvenable — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
Thank you for your work on this contribution! The code looks good and I'm approving it to merge.

I have the tests running now to validate the build passes. I'll monitor the flaky tests to be sure we get past those. Once the builds are passing, I will merge.

**Comment by dlvenable** [comment]
There are some build issues presently. I'm going to merge this PR and fix these up.

**Comment by janisbiz** [comment]
Hi @kmssap 

I am trying to figure out how to visualize histograms on OS.

Basically - is it even possible? As it is already predefined buckets (containing `min`, `max` & `count`), thus what would be the possibility to actually visualize them, as it is not possible to aggregate it properly on OS side at which point it makes this implementation completely pointless in context of OS.

Would it be possible to get some example query or smth to understand how should I aggregate e.t.c. to be able to visualize?

Thanks.

**Comment by dlvenable** [comment]
@janisbiz , I'm not familiar with the details of how they visualized the metrics.

Based on the [blog post](https://opensearch.org/blog/technical-post/2022/05/opentelemetry-metrics-for-data-prepper/) announcing this feature, I believe they used [Vega](https://vega.github.io/vega/) for visualization. I suggest you read the blog post to learn more about the end solution.
